### PR TITLE
fix(providers): analyze all FROM lines in multi-stage Dockerfiles via batch analysis

### PR DESCRIPTION
**Pull request by a-oren**
## Summary
- Fix the Dockerfile/Containerfile provider to analyze **all** FROM lines in multi-stage Dockerfiles instead of only the last one
- Each FROM image gets its own SBOM in a purl-keyed batch map routed to `/api/v5/batch-analysis`
- Resolve ARG substitutions using declared defaults (matching VS Code extension behavior)
- FROM lines with unresolvable ARGs (no default) are silently skipped

## Changes
- `src/providers/oci_dockerfile.js` — rename `parseFromImage()` → `parseAllFromImages()`, build `sbomByPurl` map, add `batch: true` flag
- `src/providers/containerfile_parser.js` — add `getArgQuery()` for ARG instruction parsing
- `src/analysis.js` — detect batch provider results and route to `requestStackBatch()`
- `test/providers/oci_dockerfile.test.js` — update tests for array return, add ARG resolution and batch output tests

Implements [TC-5070](https://redhat.atlassian.net/browse/TC-5070)

## Test plan
- [x] Unit tests: `parseAllFromImages` returns array of all image refs
- [x] Unit tests: ARG substitution resolved using declared defaults
- [x] Unit tests: Unresolvable ARGs (no default) skipped
- [x] Unit tests: Batch output format with correct purl keys
- [x] Integration tests: Dockerfile hardened recommendations scenario passes
- [x] Integration tests: Dockerfile recommendations-disabled scenario passes
- [x] `npm test` — 496 passing, 0 new failures
- [x] `npm run lint` — 0 errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[TC-5070]: https://redhat.atlassian.net/browse/TC-5070?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Analyze all FROM instructions in Dockerfiles for OCI image SBOM generation and route batch SBOM results through stack/component analysis.

New Features:
- Return SBOMs for each FROM image in multi-stage Dockerfiles as a batch map keyed by image purl.
- Support ARG instruction parsing and default-based substitution in Dockerfile FROM lines.
- Handle batch SBOM results in stack and component analysis flows via batch-analysis routing.

Enhancements:
- Rename and extend Dockerfile FROM parsing to return all resolvable base images and skip unresolvable ARG-based stages.
- Add batch flag and content handling in the Dockerfile provider and analysis pipeline to support multi-image SBOMs.

Tests:
- Expand Dockerfile provider tests to cover multi-stage parsing, ARG resolution behavior, error handling, and batch SBOM output for provideStack and provideComponent.

### CODE REVIEW

**Comment by a-oren** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review — Classified as **suggestion** (2 items):
1. De-duplicate SBOM generation for repeated images — optimization proposal, not backed by CONVENTIONS.md. No sub-task created.
2. Include unresolved variables in error message — reasonable improvement but current message is descriptive per convention. No sub-task created.

File: src/providers/containerfile_parser.js [hunk 1]
```diff
@@ -19,3 +19,8 @@ export async function getFromQuery() {
 	const language = await init();
 	return new Query(language, '(from_instruction (image_spec) @image)');
 }
+
+export async function getArgQuery() {
+	const language = await init();
+	return new Query(language, '(arg_instruction (arg_pair name: (unquoted_string) @name default: (unquoted_string) @default))');
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/providers/containerfile_parser.js]: [sdlc-workflow/verify-pr] Classified as **suggestion** — the current query correctly captures only ARGs with unquoted defaults, which is sufficient for static resolution. Broadening to wildcard node types would also require downstream normalization (quote stripping). No sub-task created.

File: test/providers/oci_dockerfile.test.js [hunk 2]
```diff
@@ -40,59 +41,99 @@ suite('testing the Dockerfile/Containerfile data provider', () => {
...
 				'FROM nginx:alpine',
 				'COPY --from=builder /app/dist /usr/share/nginx/html',
 			].join('\n')
-			expect(await parseFromImage(content)).to.equal('nginx:alpine')
+			const result = await parseAllFromImages(content)
+			expect(result).to.deep.equal(['node:18', 'nginx:alpine'])
+		})
+
+		/** Verifies that FROM lines with ARG substitution are resolved using declared defaults. */
+		test('resolves ARG substitution in FROM lines using declared defaults', async () => {
```
[the remark this answers is not in the document]
Comment (a-oren) [reply test/providers/oci_dockerfile.test.js:66]: [sdlc-workflow/verify-pr] Classified as **suggestion** — adding a test for bare $VAR syntax would improve branch coverage but is not required by CONVENTIONS.md (which sets 82% line coverage threshold, not branch-level). No sub-task created.

**Comment by a-oren** [comment]
## Verification Report for TC-5070 (commit df1dfe6)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | PASS | 4 suggestions from sourcery-ai (no code change requests) |
| Root-Cause Investigation | N/A | No sub-tasks created |
| Scope Containment | WARN | All 3 task-specified files present; `containerfile_parser.js` modified (out-of-scope but necessary for ARG query) |
| Diff Size | PASS | 308 lines across 4 files — proportionate to task scope |
| Commit Traceability | PASS | Commit `df1dfe6` references TC-5070 |
| Sensitive Patterns | PASS | No secrets or credentials detected |
| CI Status | PASS | All 5 CI checks pass (lint/test Node 22/24, Sourcery, PR title, commit messages) |
| Acceptance Criteria | PASS | All 6 criteria satisfied — multi-stage FROM parsing, batch SBOM map, batch routing, single-FROM, ARG resolution, Dockerfile/Containerfile support |
| Test Quality | WARN | Test Documentation: PASS (all tests have JSDoc); Repetitive Test Detection: WARN (some parseAllFromImages tests share identical structure); Eval Quality: N/A |
| Test Change Classification | ADDITIVE | 8 net-new tests added, 0 removed; coverage increased from 90.75% to 90.99% |
| Verification Commands | PASS | All 33 Dockerfile/Containerfile tests pass; lint has 0 errors |

### Overall: WARN

**Scope Containment details:**

`src/providers/containerfile_parser.js` was modified (+5 lines) to add `getArgQuery()` for ARG instruction parsing. This file is not listed in the task spec's "Files to Modify" but is a tightly-coupled dependency required for the ARG substitution feature described in the Implementation Notes.

**Repetitive Test Detection details:**

Several `parseAllFromImages` tests (single FROM, platform flag, digest, case-insensitive FROM, single FROM with ARG) share identical structure: define content string → call `parseAllFromImages(content)` → assert `deep.equal` on result array. These could potentially be parameterized using a data-driven pattern, though individually they test distinct parsing edge cases.

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.12.2.*

File: src/providers/containerfile_parser.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Strum355) [reply src/providers/containerfile_parser.js]: ~~Although Ive never seen quoted ones in the wild~~ (I was thinking of something else), we should ideally handle them. Removing the quotes will be necessary

File: src/providers/oci_dockerfile.js [hunk 3]
```diff
@@ -51,43 +51,103 @@ function containsExpansion(node) {
...
 async function getImageSBOM(manifest, opts = {}) {
 	const manifestContent = fs.readFileSync(manifest, 'utf-8')
-	const image = await parseFromImage(manifestContent)
-	const imageRef = parseImageRef(image, opts)
-	const sbom = generateImageSBOM(imageRef, opts)
+	const images = await parseAllFromImages(manifestContent)
+	const sbomByPurl = {}
+	for (const image of images) {
+		const imageRef = parseImageRef(image, opts)
+		sbomByPurl[imageRef.getPackageURL().toString()] = generateImageSBOM(imageRef, opts)
```
Comment (Strum355) [inline src/providers/oci_dockerfile.js]: If a containerfile lists the same image multiple times, this is rerun unnecessarily

**Review by Strum355 — DISMISSED** [review]
Some small changes but LGTM otherwise

**Review by Strum355 — APPROVED** [review]
